hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/action/index/HoodieRecordBasedIndexingCatchup.java

```diff
@@ -0,0 +1,156 @@
+      // metadata index already updated for this instant
+      instant = getHoodieInstant(instant);
+      if (instant == null) {
+        continue;
+      }
+      // if instant completed, ensure that there was metadata commit, else update metadata for this completed instant
+      if (COMPLETED.equals(instant.getState())) {
+        String instantTime = instant.getTimestamp();
+        Option<HoodieInstant> metadataInstant = metadataMetaClient.reloadActiveTimeline()
+            .filterCompletedInstants().filter(i -> i.getTimestamp().equals(instantTime)).firstInstant();
+        if (metadataInstant.isPresent()) {
+          currentCaughtupInstant = instantTime;
+          continue;
+        }
+        try {
+          // we need take a lock here as inflight writer could also try to update the timeline
+          transactionManager.beginTransaction(Option.of(instant), Option.empty());
+          LOG.info("Updating metadata table for instant: " + instant);
+          switch (instant.getAction()) {
+            case HoodieTimeline.COMMIT_ACTION:
+            case HoodieTimeline.DELTA_COMMIT_ACTION:
+            case HoodieTimeline.REPLACE_COMMIT_ACTION:
+              HoodieCommitMetadata commitMetadata = HoodieCommitMetadata.fromBytes(
+                  metaClient.getActiveTimeline().getInstantDetails(instant).get(), HoodieCommitMetadata.class);
+              HoodieData<HoodieRecord> records = readRecordsFromDataFiles(instant);
+              metadataWriter.update(commitMetadata, records, instant.getTimestamp());
+              break;
+            case CLEAN_ACTION:
+              HoodieCleanMetadata cleanMetadata = CleanerUtils.getCleanerMetadata(metaClient, instant);
+              metadataWriter.update(cleanMetadata, instant.getTimestamp());
+              break;
+            case RESTORE_ACTION:
+              HoodieRestoreMetadata restoreMetadata = TimelineMetadataUtils.deserializeHoodieRestoreMetadata(
+                  metaClient.getActiveTimeline().getInstantDetails(instant).get());
+              metadataWriter.update(restoreMetadata, instant.getTimestamp());
+              break;
+            case ROLLBACK_ACTION:
+              HoodieRollbackMetadata rollbackMetadata = TimelineMetadataUtils.deserializeHoodieRollbackMetadata(
+                  metaClient.getActiveTimeline().getInstantDetails(instant).get());
+              metadataWriter.update(rollbackMetadata, instant.getTimestamp());
+              break;
+            default:
+              throw new IllegalStateException("Unexpected value: " + instant.getAction());
+          }
+        } catch (IOException e) {
+          throw new HoodieIndexException(String.format("Could not update metadata partition for instant: %s", instant), e);
+        } finally {
+          transactionManager.endTransaction(Option.of(instant));
+        }
+      }
+    }
+  }
+
+  private HoodieData<HoodieRecord> readRecordsFromDataFiles(HoodieInstant instant) throws IOException {
```

Comment:
Base vs Data? Inside the method, we are looking at the base file only?

Reply:
renamed to denote we're rading from file slices and not base files.